iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -10,568 +10,728 @@
+    char* return_value = (char*)_malloc_and_copy_unsigned_char(value, length);
+
+    json_value_free(root_value);
+
+    return return_value;
+}
+
+static int _parse_json_twin_int_from_array(const char* twin_payload,
+                                           const char* full_property_name,
+                                           size_t index,
+                                           bool allow_for_zero)
+{
+    JSON_Value* root_value = json_parse_string(twin_payload);
+    ASSERT_IS_NOT_NULL(root_value);
+    JSON_Object* root_object = json_value_get_object(root_value);
+    ASSERT_IS_NOT_NULL(root_object);
+
+    JSON_Array* array = json_object_dotget_array(root_object, full_property_name);
+    ASSERT_IS_NOT_NULL(array, "Array not specified.");
+
+    double double_value = json_array_get_number(array, index);
+    int int_value = (int)(double_value + 0.1); // Account for possible underflow by small increment.
+
+    if (!allow_for_zero)
     {
-        result = IoTHubDeviceClient_SetDeviceTwinCallback(iothub_deviceclient_handle, twinCallback, device);
+        ASSERT_ARE_NOT_EQUAL(int, 0, int_value, "Failed to parse `%s`.", full_property_name);
     }
 
-    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result, "IoTHub(Device|Module)Client_SetDeviceTwinCallback failed");
+    json_value_free(root_value);
+
+    return int_value;
 }
 
-static void sendeventasync_on_device_or_module(IOTHUB_MESSAGE_HANDLE msgHandle)
+//
+// Service Client APIs
+//
+static void _service_client_update_twin(IOTHUB_SERVICE_CLIENT_DEVICE_TWIN_HANDLE serviceclient_devicetwin_handle,
+                                        IOTHUB_PROVISIONED_DEVICE* device_to_use,
+                                        const char* twin_json)
 {
-    IOTHUB_CLIENT_RESULT result;
+    char* twin_response;
 
-    if (iothub_moduleclient_handle != NULL)
+    LogInfo("_service_client_update_twin(): Beginning update of twin via Service SDK.");
+
+    if (device_to_use->moduleId != NULL)
     {
-        result = IoTHubModuleClient_SendEventAsync(iothub_moduleclient_handle, msgHandle, NULL, NULL);
+        twin_response = IoTHubDeviceTwin_UpdateModuleTwin(serviceclient_devicetwin_handle,
+                                                          device_to_use->deviceId,
+                                                          device_to_use->moduleId, twin_json);
     }
     else
     {
-        result = IoTHubDeviceClient_SendEventAsync(iothub_deviceclient_handle, msgHandle, NULL, NULL);
+        twin_response = IoTHubDeviceTwin_UpdateTwin(serviceclient_devicetwin_handle,
+                                                    device_to_use->deviceId, twin_json);
     }
 
-    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result, "SendEventAsync failed");
-}
-
+    ASSERT_IS_NOT_NULL(twin_response, "IoTHubDeviceTwin_Update(Module)Twin failed.");
 
+    LogInfo("_service_client_update_twin(): Twin response from Service SDK after update is <%s>.\n",
```

Comment:
Please remove the "\n" in LogInfo string - by default it'll concat this for us and most other places don't do this.  I understand some of legacy code in this did the \n too but it should get cleaned up too.